client/output.go

```diff
@@ -58,37 +56,45 @@ func (c *Client) GetOutput(ctx context.Context, buildID string, or OutputReader)
 	defer resp.Body.Close()
 	defer ws.Close()
 
-	go func() {
-		sigCh := make(chan os.Signal, 1)
-		signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM)
-
-		fmt.Printf("\rShutting down due to signal: %v\n", <-sigCh)
-
-		if err := c.Cancel(ctx, buildID); err != nil {
-			c.Logger.Logf("build cancellation request failed: %v", err)
-		}
+	errChan := make(chan error)
 
-		cancel()
+	go func() {
+		defer close(errChan)
+		errChan <- func() error {
+			for {
+				// Read from websocket
+				mt, msg, err := ws.ReadMessage()
+				if err != nil {
+					if websocket.IsCloseError(err, websocket.CloseNormalClosure) {
+						return nil
+					}
+					c.Logger.Logf("websocket read message err - %s", err)
+					return err
+				}
+
+				n, err := or.Read(mt, msg)
+				if err != nil {
+					return err
+				}
+				if n != len(msg) {
+					return fmt.Errorf("did not read all message contents: %d != %d", n, len(msg))
+				}
+			}
+		}()
 	}()
 
-	for {
-		// Read from websocket
-		mt, msg, err := ws.ReadMessage()
-		if err != nil {
-			if websocket.IsCloseError(err, websocket.CloseNormalClosure) {
-				return nil
-			}
-			c.Logger.Logf("websocket read message err - %s", err)
-			return err
-		}
+	select {
+	case <-ctx.Done():
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
 
-		n, err := or.Read(mt, msg)
-		if err != nil {
-			return err
-		}
-		if n != len(msg) {
-			return fmt.Errorf("did not read all message contents: %d != %d", n, len(msg))
+		if err := c.Cancel(ctx, buildID); err != nil { //nolint
+			c.Logger.Logf("build cancellation request failed: %v", err)
 		}
 
+		<-errChan
```

Comment:
Concerned this could hang for a very long time if `c.Cancel` fails, or the server doesn't behave as expected. In my view, our responsibility on context cancellation is to send a cancel request to the server, and tear down the client ASAP.

Along those lines, I'm wondering if we should close `ws` here to minimize the chances this blocks:

```suggestion
		ws.Close()

		<-errChan
```

Reply:
I think this makes sense!